### ISSUE TITLE
Free method of obtaining xmltv file

### ISSUE THREAD

**Issue by deathbybandaid** [tgorgdotcom/locast2plex#66]
Not an issue, just something that I just finished working on, and may be of use to other locast2plex users.

I modified the script from this repo to pull the xmltv file, and then I serve it via nginx.

https://github.com/arantius/zap2xml-py/blob/master/zap2xml.py

My version:

https://pastebin.pl/view/533a579b

I modified this to also retrieve multiple timezones, as well as the poster/channel thumbnails.

**Comment by tgorgdotcom** [comment]
Just to add, Locast has EPG data we can grab -- I haven't had time to look into it, but I think the data comes with the request to https://api.locastnet.org/api/watch/epg/

**Comment by java007md** [comment]
**tgordotcom:** Obtaining the EPG data through Locast would be ideal. The code I am trying to use for importing xmltv data is from:

https://github.com/sjsaleem/zap2xml

Seems to be working for the most part for populating the guide in Jellyfin, but I am still working through some issues with the tuner setup in general. Will have to give the code from **deathbybandaid** a try. Thanks!

**Comment by deathbybandaid** [comment]
@java007md If you try out my thing, use this instead of the pastebin, I did some additional cleanup.

The only thing I can't seem to get from Zap is channel names like `CBS` instead of just the callsign.

I also consulted documentation on the xmltv standard to try and cleanup the xml produced by this.

I also added a `--output_file` argument so that I can run this for 2 zipcodes.

https://github.com/deathbybandaid/zap2xml-py/blob/master/zap2xml.py

**Comment by deathbybandaid** [comment]
@tgorgdotcom looks like you are very much right. The only thing it needs is the DMA number added

https://api.locastnet.org/api/watch/epg/501

looks like very easy information to parse, I'll take a crack at it in a seperate repo based on the zap one linked above

**Comment by deathbybandaid** [comment]
@tgorgdotcom okay, so here's the basic concept for converting the epg to xmltv format

**Comment by jcastilloalonso** [comment]
> 
> 
> @tgorgdotcom okay, so here's the basic concept for converting the epg to xmltv format
> 
> https://github.com/deathbybandaid/locastepg

Nice! Wonder if making a pull request would be better so it can be easier merged into the project?

My only question is why would it be better to use the epg/xmltv format vs the integrated Plex guide? So far I have yet to see any problems, but I'm sure there are valid scenarios.

**Comment by deathbybandaid** [comment]
There's a PR I put in regarding a code cleanup that I'd like to see beforehand, so I can base on that. The seperate repo was me just testing, and I fully plan to rework it a bit more for direct integration. Might also keep the seperate repo, as it could be useful in other ways.

Locast2plex also work with other players (anything that can pull from a 'hdhr'), so having the guide data built in here is a nice convenience.

**Comment by deathbybandaid** [comment]
Side note, for implementation here, we might consider having both xmltv.xml as well as epg.json accessible since we pull it anyway.

**Comment by deathbybandaid** [comment]
my new method apparently needs work, need to figure out the time formatting that I need.

Locast only presents Start time, and in Epoch, but with milliseconds.
Locast does provide duration, but in seconds, so End time is calculated from these pieces of information.

**Comment by deathbybandaid** [comment]
Okay, this morning, with a clear mind, I managed to get the timestamps working properly, and my guide now loads. Turns out the date format is very specific for xmltv

**Comment by deathbybandaid** [comment]
Made more progress with this, and thanks to some help from @thlayli123 I now am able to retrieve 7 days worth of EPG.

I'm waiting on the switch to Py3 to port that code into locast2plex.

**Comment by mikeh143** [comment]
Is there any way to also output an m3u file? I'd like to integrate locast2plex into xteve, but it asks for an m3u

**Comment by deathbybandaid** [comment]
@mikeh143 add locast2plex to xteve as a HDHomeRun.

I have a PR in the works to add xmltv directly into locast2plex.

**Comment by eqpaisley** [comment]
Hi all, is there anymore working being done on this. A .XML file for Locast would be incredibly helpful, obviously it would have to be updated on a regular basis.

**Comment by teleflix** [comment]
Hello

but about locast epg.data is not compatible on so many popular iptv pannels? for use it requeired convert from epg data to xmlt right? ot some idea for this epg data wotk on xtream-ui for example?

thank
